### ISSUE TITLE
Add asArray() and asObject() to JsonNode

### ISSUE THREAD

**Issue by victornoel** [FasterXML/jackson-databind#2343]
In the same way that there is various methods to manipulate a `JsonNode` to one of its potential real type (String, boolean, etc), there should be `asArray()` and `asObject()` available.

Currently the only way to retrieve a `JsonNode` as an `ObjectNode` or an `ArrayNode` is with the two methods `with(String)` and `withArray(String)` (even though in 2.9 their return type is broken).

But for example when manipulating an `ArrayNode`, it is not possible to get its elements as `ArrayNode` or `ObjectNode` (or I didn't find the way to do it).

I would expect to be able, for example, to do:
```java
for(JsonNode node: anArrayNode) {
  node.asObject().set("a", new TextNode("b"));
}
```

**Comment by cowtowncoder** [comment]
Makes sense. I think am including that as part of bigger package of improvements for `JsonNode`:

https://github.com/FasterXML/jackson-future-ideas/wiki/JSTEP-3

and the only (?) question there would probably be what to do with non-existing value (whether to throw exception or return `MissingNode`). This is bit different from `with()`, which automatically creates an entry. I think I'd go with `MissingNode`, and then have "required" variant that throws exception.

**Comment by cowtowncoder** [comment]
Oh. Actually, returning `MissingNode` makes no sense as result type would be expected to be `ArrayNode` / `ObjectNode`, respectively. So exception needed.

**Comment by cowtowncoder** [comment]
Implemented: also added:

* `Optional<ArrayNode> JsonNode.asArrayOpt()`
* `Optional<ObjectNode> JsonNode.asObjectOpt()`
